### PR TITLE
fix zettacache locking

### PR DESCRIPTION
**Pull request by ahrens**
The zettacache buffers insertion requests in memory, up to a certain
limit, writing them to disk as the disk can keep up.  Once an insertion
is assigned to a specific checkpoint (which is always "the current
checkpoint" - there's no open/quiescing/syncing TXG train), that
checkpoint needs to wait for the insertion to complete (including its
write to disk) before the checkpoint can be completed.  This ensures
that if we crash, the logical index (typically the operation log)
doesn't refer to disk locations that weren't actually written.

The problem is that we assign an insertion to the checkpoint early -
when it's buffered.  (This happens when we add it to
ZettaCache::outstanding_inserts.)  It may remain buffered for a long
time (dozens of seconds), depending on CACHE_INSERT_MAX_BUFFER and
underlying disk performance.

When writing out a checkpoint, we have the ZettaCacheState lock, and we
wait for the insertions assigned to this checkpoint to complete (the
outstanding_inserts).  If this takes a long time, it can be very bad for
performance because new lookups and insertions have to wait (many
seconds) for the zettacache state lock.

The solution implemented in this PR is to wait until later to assign
insertions to a checkpoint, waiting until we are able to issue a write
to disk.

As part of a separate PR, we may want to reduce the impact of the (now
small, ~1 second) time it takes to sync a checkpoint.  For example by
having lookup/insert requests fail rather than wait for the state lock.
Or to have a open/syncing train like we do for TXG's, where we can
accumulate more changes in "open context" while we're in the middle of
writing out a batch of changes from "syncing context".

### CODE REVIEW

File: cmd/zfs_object_agent/zettacache/src/block_access.rs [hunk 1]
```diff
@@ -18,16 +18,21 @@ use std::convert::TryFrom;
...
 use tokio::sync::Semaphore;
+use tokio::task::JoinHandle;
 use util::get_tunable;
 use util::From64;
 
 lazy_static! {
     static ref MIN_SECTOR_SIZE: usize = get_tunable("min_sector_size", 512);
     static ref DISK_WRITE_MAX_QUEUE_DEPTH: usize = get_tunable("disk_write_max_queue_depth", 32);
+    static ref DISK_WRITE_METADATA_MAX_QUEUE_DEPTH: usize =
+        get_tunable("disk_write_max_queue_depth", 32);
```
Comment (mmaybee) [inline cmd/zfs_object_agent/zettacache/src/block_access.rs:35]: You are using the same tunable name here twice: `disk_write_max_queue_depth`, is that intentional?

File: cmd/zfs_object_agent/zettacache/src/zettacache.rs [hunk 2]
```diff
@@ -1033,13 +1032,32 @@ impl ZettaCache {
             Err(e) => panic!("unexpected error from try_acquire: {:?}", e),
         };
 
-        let mut state = self.state.lock().await;
-        let index_key = locked_key.0.value();
-        state.insert(permit, index_key.guid, index_key.block, buf);
+        let block_access = self.block_access.clone();
+        let state = self.state.clone();
+        tokio::spawn(async move {
```
Comment (mmaybee) [inline cmd/zfs_object_agent/zettacache/src/zettacache.rs:1037]: So, if I am understanding this correctly, you are spawning a task for each insert. This task waits (potentially a long time) for the write permit (is this the time in the queue?) and only after it has this permit will it acquire the state lock to do the actual allocate/write.
Are there any scaling issues with large numbers of outstanding tasks?

File: cmd/zfs_object_agent/zettacache/src/block_access.rs (hunk 1 above)
Comment (ahrens) [reply cmd/zfs_object_agent/zettacache/src/block_access.rs:35]: nope, copy/paste error, thanks for pointing that out!

File: cmd/zfs_object_agent/zettacache/src/zettacache.rs (hunk 2 above)
Comment (ahrens) [reply cmd/zfs_object_agent/zettacache/src/zettacache.rs:1037]: Correct.  During the potentially long wait, we have the LockedKey and the insert_permit, and are waiting for the WritePermit.  So we are preventing concurrent lookups/inserts of this key (PoolGUID+BlockId), and the insert_permit tracks how much memory we are using for the buffer (roughly - need to change this from number of blocks to number of bytes).

As to scaling issues:
- I've read that millions of tasks are no problem (unverified)
- In practice I've seen nontrivial CPU used for scheduling/descheduling tasks, but I don't know if that's related to number of tasks, vs just frequency of going on/off thread
- This PR doesn't really change any of this; the old code was spawning a task to do the write in ZettaCacheState::insert, and I've removed that spawn and replaced it with this one in ZettaCache::insert.  So we have the same number of spawns as before (one "normal" plus one "blocking" (see BlockAccess::write_raw_permit()) per insert).

Comment (mmaybee) [reply cmd/zfs_object_agent/zettacache/src/zettacache.rs:1037]: Makes sense.

**Review by mmaybee — APPROVED** [review]

**Pull request merged**